### PR TITLE
Hive file format cleanup

### PR DESCRIPTION
**Pull request by electrum**

### CODE REVIEW

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java [hunk 1]
```diff
@@ -3685,7 +3671,7 @@ private void assertOneNotNullResult(@Language("SQL") String query)
 
     private boolean insertOperationsSupported(HiveStorageFormat storageFormat)
     {
-        return storageFormat != HiveStorageFormat.DWRF;
+        return true;
```
Comment (findepi) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java]: remove `insertOperationsSupported` method?

File: presto-orc/src/main/java/io/prestosql/orc/OrcEncoding.java [hunk 2]
```diff
@@ -34,19 +32,6 @@ public MetadataWriter createMetadataWriter()
         {
             return new OrcMetadataWriter();
         }
-    },
-    DWRF {
```
Comment (findepi) [inline presto-orc/src/main/java/io/prestosql/orc/OrcEncoding.java]: remove `OrcEncoding` enum

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HiveClientConfig.java [hunk 3]
```diff
@@ -800,20 +801,6 @@ public HiveClientConfig setOrcDefaultBloomFilterFpp(double orcDefaultBloomFilter
...
     }
 
-    @Deprecated
-    public boolean isOrcOptimizedWriterEnabled()
-    {
-        return orcOptimizedWriterEnabled;
-    }
-
-    @Deprecated
-    @Config("hive.orc.optimized-writer.enabled")
```
Comment (findepi) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/HiveClientConfig.java]: When removing old writer, we need a new toggle, that would make the new writer "fake" writer version (and writer id).

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java [hunk 4]
```diff
@@ -2894,19 +2890,14 @@ public void testRcTextCharDecoding()
 
     private void testRcTextCharDecoding(boolean rcFileOptimizedWriterEnabled)
```
Comment (findepi) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java]: remove `rcFileOptimizedWriterEnabled` param

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java (hunk 1 above)
Comment (electrum) [reply presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java]: Thanks, missed that

File: presto-orc/src/main/java/io/prestosql/orc/OrcEncoding.java (hunk 2 above)
Comment (electrum) [reply presto-orc/src/main/java/io/prestosql/orc/OrcEncoding.java]: Also missed

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java (hunk 4 above)
Comment (electrum) [reply presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java]: Done

**Comment by electrum** [comment]
The legacy writers use the shared `RecordFileWriter`. Your comment reminded me of a couple additional things:

* `createRcFileWriter` in `HiveWriteUtils`
* `OptimizedLazyBinaryColumnarSerde` usage in `RecordFileWriter` (we can remove this from the shaded Hive build)

For DWRF, there might be further simplifications, such as removal of no longer needed abstractions, that can be done in the future.

**Comment by electrum** [comment]
I added another commit that allows writing compatible ORC files.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HiveClientConfig.java [hunk 5]
```diff
@@ -801,6 +802,19 @@ public HiveClientConfig setOrcDefaultBloomFilterFpp(double orcDefaultBloomFilter
         return this;
     }
 
+    public boolean isOrcWriteLegacyVersion()
+    {
+        return orcWriteLegacyVersion;
+    }
+
+    @Config("hive.orc.writer.use-legacy-version-number")
+    @ConfigDescription("Use a writer version number that is readable by early Hive 2.x releases")
```
Comment (dain) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/HiveClientConfig.java]: Should this be "2.3 and earlier releases"?

File: presto-orc/src/main/java/io/prestosql/orc/metadata/OrcMetadataWriter.java [hunk 6]
```diff
@@ -98,7 +108,7 @@ public int writeFooter(SliceOutput output, Footer footer)
             throws IOException
     {
         OrcProto.Footer footerProtobuf = OrcProto.Footer.newBuilder()
-                .setWriter(PRESTO_WRITER_ID)
+                .setWriter(useLegacyVersion ? HIVE_WRITER_ID : PRESTO_WRITER_ID)
```
Comment (findepi) [inline presto-orc/src/main/java/io/prestosql/orc/metadata/OrcMetadataWriter.java]: Maybe in legacy we shouldn't write writer id at all?

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HiveClientConfig.java (hunk 5 above)
Comment (findepi) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/HiveClientConfig.java]: @dain i agree; especially that I _think_ problem existed in Hive 1.2 as well 

@electrum "*early* Hive 2.x releases" seems to imply "you are unlikely to need this", whereas currently _every_ Cloudera customer is affected, as their _latest_ distro still ships Hive 2.1.1 (AFAIK CDH 6.1.1 is the latest)

File: presto-orc/src/main/java/io/prestosql/orc/metadata/OrcMetadataWriter.java (hunk 6 above)
Comment (electrum) [reply presto-orc/src/main/java/io/prestosql/orc/metadata/OrcMetadataWriter.java]: Good idea. That field wasn't added until writer version 6.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HiveClientConfig.java (hunk 5 above)
Comment (electrum) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/HiveClientConfig.java]: I confirmed that the files are readable by Hive 1.2.1, so it's only Hive 2.x, but I agree that this should be more explicit. I'll change it to

> Write ORC files with a version number that is readable by Hive 2.0.0 to 2.2.0

**Pull request merged**